cpp/src/interop/from_arrow.cu

```diff
@@ -104,16 +104,22 @@ struct dispatch_to_cudf_column {
     if (array.null_bitmap_data() == nullptr) {
       return std::make_unique<rmm::device_buffer>(0, stream, mr);
     }
-    auto mask = std::make_unique<rmm::device_buffer>(
-      bitmask_allocation_size_bytes(static_cast<size_type>(array.null_bitmap()->size() * CHAR_BIT)),
-      stream,
-      mr);
+    auto const null_bitmap_size = array.null_bitmap()->size();
+    auto const allocation_size =
+      bitmask_allocation_size_bytes(static_cast<size_type>(null_bitmap_size * CHAR_BIT));
+    auto mask        = std::make_unique<rmm::device_buffer>(allocation_size, stream, mr);
     auto mask_buffer = array.null_bitmap();
     CUDF_CUDA_TRY(cudaMemcpyAsync(mask->data(),
                                   reinterpret_cast<const uint8_t*>(mask_buffer->address()),
-                                  array.null_bitmap()->size(),
+                                  null_bitmap_size,
                                   cudaMemcpyDefault,
                                   stream.value()));
+    // Zero-initialize trailing padding bytes
+    auto const num_trailing_bytes = allocation_size - null_bitmap_size;
+    if (num_trailing_bytes > 0) {
+      auto trailing_bytes = static_cast<uint8_t*>(mask->data()) + null_bitmap_size;
+      CUDF_CUDA_TRY(cudaMemsetAsync(trailing_bytes, 0, num_trailing_bytes, stream.value()));
```

Comment:
Considering this extra CUDA call is being added to a stack trace that includes calls to `null_count()` I think it we should evaluate the performance impact of this change across cudf.

Reply:
Happy to do the evaluation, although I thought this would only be called during external data ingest from an arrow table. So once we've got a `column` with an allocated mask we won't touch this codepath?

Reply:
Ok, I wasn't sure how much this is called by the cudf layer. I trust your evaluation of the impact.

Reply:
Here's a comparison of trunk and this branch for the gbench benchmarks. Everything looks to be within the noise window.

I ran:

```sh
for f in branch-23.04/release/benchmarks/*_BENCH; do
    python ~/third-party/google-benchmark/tools/compare.py --no-color benchmarks \
           branch-23.04/release/benchmarks/$(basename $f)_97a746f.json \
           wence__fix__from-arrow-bitmask-initialization/release/benchmarks/$(basename $f)_1dc9e64.json
    printf "\n\n"
done | tee benchmark-comparison.txt
```

[benchmark-comparison.txt](https://github.com/rapidsai/cudf/files/10668088/benchmark-comparison.txt)

Summary (passing `-a` as well) of the results:

```
Comparing branch-23.04/release/benchmarks/APPLY_BOOLEAN_MASK_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/APPLY_BOOLEAN_MASK_BENCH_1dc9e64.json
Benchmark                                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
------------------------------------------------------------------------------------------------------------------------------------------------
ApplyBooleanMask<float>/float_1_col/100000000/50_pvalue           0.6667          0.6667      U Test, Repetitions: 2 vs 2. WARNING: Results unreliable! 9+ repetitions recommended.
ApplyBooleanMask<float>/float_2_col/100000000/50_pvalue           0.6667          0.6667      U Test, Repetitions: 2 vs 2. WARNING: Results unreliable! 9+ repetitions recommended.
ApplyBooleanMask<float>/float_4_col/100000000/50_pvalue           0.3333          0.6667      U Test, Repetitions: 2 vs 2. WARNING: Results unreliable! 9+ repetitions recommended.
OVERALL_GEOMEAN                                                  +0.0015         +0.0014             0             0             0             0


Comparing branch-23.04/release/benchmarks/AST_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/AST_BENCH_1dc9e64.json
Benchmark                                                                                                                         Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                                                                                -0.0010         -0.0020             0             0             0             0


Comparing branch-23.04/release/benchmarks/BINARYOP_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/BINARYOP_BENCH_1dc9e64.json
Benchmark                                                                                                                       Time             CPU      Time Old      Time New       CPU Old       CPU New
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                                                                              +0.0009         +0.0011             0             0             0             0


Comparing branch-23.04/release/benchmarks/COLUMN_CONCAT_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/COLUMN_CONCAT_BENCH_1dc9e64.json
Benchmark                                                                                     Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                                            +0.0056         +0.0050             0             0             0             0


Comparing branch-23.04/release/benchmarks/CONTIGUOUS_SPLIT_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/CONTIGUOUS_SPLIT_BENCH_1dc9e64.json
Benchmark                                                                                                   Time             CPU      Time Old      Time New       CPU Old       CPU New
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                                                          +0.0071         +0.0072             0             0             0             0


Comparing branch-23.04/release/benchmarks/COPY_IF_ELSE_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/COPY_IF_ELSE_BENCH_1dc9e64.json
Benchmark                                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                  +0.0185         +0.0066             0             0             0             0


Comparing branch-23.04/release/benchmarks/CSV_WRITER_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/CSV_WRITER_BENCH_1dc9e64.json
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                   +0.0016         +0.0003             1             1             1             1


Comparing branch-23.04/release/benchmarks/FILL_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/FILL_BENCH_1dc9e64.json
Benchmark                                                       Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                              +0.0032         +0.0037             0             0             0             0


Comparing branch-23.04/release/benchmarks/GATHER_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/GATHER_BENCH_1dc9e64.json
Benchmark                                                         Time             CPU      Time Old      Time New       CPU Old       CPU New
----------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                -0.0009         +0.0002             0             0             0             0


Comparing branch-23.04/release/benchmarks/GROUPBY_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/GROUPBY_BENCH_1dc9e64.json
Benchmark                                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                  +0.0028         +0.0026             0             0             0             0


Comparing branch-23.04/release/benchmarks/HASHING_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/HASHING_BENCH_1dc9e64.json
Benchmark                                                                        Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                               +0.0067         +0.0061             0             0             0             0


Comparing branch-23.04/release/benchmarks/ITERATOR_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/ITERATOR_BENCH_1dc9e64.json
Benchmark                                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                 +0.0325         +0.0228             0             0             0             0


Comparing branch-23.04/release/benchmarks/JOIN_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/JOIN_BENCH_1dc9e64.json
Benchmark                                                                                                             Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                                                                    +0.0013         +0.0015             0             0             0             0


Comparing branch-23.04/release/benchmarks/JSON_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/JSON_BENCH_1dc9e64.json
Benchmark                                                 Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                        -0.0211         -0.0201             0             0             0             0


Comparing branch-23.04/release/benchmarks/MERGE_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/MERGE_BENCH_1dc9e64.json
Benchmark                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                 +0.0032         +0.0031             0             0             0             0


Comparing branch-23.04/release/benchmarks/NULLMASK_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/NULLMASK_BENCH_1dc9e64.json
Benchmark                                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
---------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                     -0.0180         -0.0003             0             0             0             0


Comparing branch-23.04/release/benchmarks/QUANTILES_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/QUANTILES_BENCH_1dc9e64.json
Benchmark                                                      Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                             +0.0013         +0.0012             0             0             0             0


Comparing branch-23.04/release/benchmarks/REDUCTION_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/REDUCTION_BENCH_1dc9e64.json
Benchmark                                                                Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                       -0.0103         -0.0058             0             0             0             0


Comparing branch-23.04/release/benchmarks/REPLACE_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/REPLACE_BENCH_1dc9e64.json
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                   -0.0072         -0.0060             0             0             0             0


Comparing branch-23.04/release/benchmarks/SCATTER_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/SCATTER_BENCH_1dc9e64.json
Benchmark                                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                 +0.0072         +0.0052             0             0             0             0


Comparing branch-23.04/release/benchmarks/SCATTER_LISTS_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/SCATTER_LISTS_BENCH_1dc9e64.json
Benchmark                                                                      Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                             -0.0012         -0.0010             0             0             0             0


Comparing branch-23.04/release/benchmarks/SEARCH_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/SEARCH_BENCH_1dc9e64.json
Benchmark                                                             Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                    +0.0013         +0.0013             0             0             0             0


Comparing branch-23.04/release/benchmarks/SHIFT_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/SHIFT_BENCH_1dc9e64.json
Benchmark                                                                     Time             CPU      Time Old      Time New       CPU Old       CPU New
----------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                            +0.0056         +0.0024             0             0             0             0


Comparing branch-23.04/release/benchmarks/SORT_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/SORT_BENCH_1dc9e64.json
Benchmark                                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
---------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                     -0.0056         -0.0055             0             0             0             0


Comparing branch-23.04/release/benchmarks/STRINGS_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/STRINGS_BENCH_1dc9e64.json
Benchmark                                                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
----------------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                                  -0.0022         -0.0017             0             0             0             0


Comparing branch-23.04/release/benchmarks/TEXT_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/TEXT_BENCH_1dc9e64.json
Benchmark                                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                 -0.0004         -0.0006             0             0             0             0


Comparing branch-23.04/release/benchmarks/TRANSPOSE_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/TRANSPOSE_BENCH_1dc9e64.json
Benchmark                                                      Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                             -0.0108         -0.0087             0             0             0             0


Comparing branch-23.04/release/benchmarks/TYPE_DISPATCHER_BENCH_97a746f.json to wence__fix__from-arrow-bitmask-initialization/release/benchmarks/TYPE_DISPATCHER_BENCH_1dc9e64.json
Benchmark                                                                       Time             CPU      Time Old      Time New       CPU Old       CPU New
------------------------------------------------------------------------------------------------------------------------------------------------------------
OVERALL_GEOMEAN                                                              -0.0043         -0.0014             0             0             0             0


```